lib/CircuitBreaker.js

```diff
@@ -1,10 +1,150 @@
 /* eslint no-param-reassign: 0 */
+/* eslint no-template-curly-in-string: 0 */
 
 const { ZenkoMetrics } = require('arsenal').metrics;
 
 const collectDefaultMetricsIntervalMs = 10000;
 
+/**
+ * circuit breaker configuration adapted for the bucket processor
+ * @typedef {Object} BucketProcessorCircuitBreakerConfig
+ * @property {Object[]} global - global probe configurations (circuit breaks everything)
+ * @property {Object.<string, Object[]>} topic - Per topic probe configurations
+ * @property {Object.<string, Object[]>} location - Per location probe configurations
+ */
+
+/**
+ * returns a list of relevant topics to the bucket processor
+ * @param {Object} lcConfig lifecycle configuration
+ * @param {Object} repConfig replication configuration
+ * @param {Object} locations  location configuration
+ * @returns {string[]} list of topics
+ */
+function getAllTopics(lcConfig, repConfig, locations) {
+    const topics = [
+        lcConfig.objectTasksTopic,
+        repConfig.dataMoverTopic,
+    ];
+    Object.keys(locations).forEach(locationName => {
+        if (locations[locationName].isCold) {
+            topics.push(lcConfig.coldStorageArchiveTopicPrefix + locationName);
+        }
+    });
+    return topics;
+}
+
+/**
+ * Adapts the templated circuit breaker configuration to the bucket processor
+ * Invalid probes are ignored
+ * @param {Object} cbConf templated circuit breaker configuration
+ * @param {Object} lcConfig lifecycle configuration
+ * @param {Object} repConfig replication configuration
+ * @param {Object} locations location configuration
+ * @returns {BucketProcessorCircuitBreakerConfig} circuit breaker configuration
+ */
+function extractBucketProcessorCircuitBreakerConfigs(cbConf, lcConfig, repConfig, locations) {
+    const configs = {
+        global: [],
+        topic: {},
+        location: {},
+    };
+
+    if (!cbConf || !cbConf.probes) {
+        return configs;
+    }
+
+    cbConf.probes.forEach(probe => {
+        if (probe.type !== 'prometheusQuery') {
+            configs.global.push(probe);
+            return;
+        }
+        // with clause is optional and can be used to specify the filtering on
+        // the topic and location, along with specifying if a probe should be
+        // used for transition and expiration.
+        // Query syntax:
+        // "with(topic=topicName,location=locationName,transition=true,expiration=true) and
+        //      PROM_METRIC_NAME{topic='${topic}',location='${location}'}"
+        // Note: should either specify the topic or the location, not both
+        // Behaviour:
+        // - no with clause + templated topic: creates a circuit breaker for each topic
+        // - with clause with a specific topic + templated topic: creates a circuit breaker for the specified topic
+        // - with clause with a specific location + templated location: creates a circuit breaker for the location
+        const withClause = probe.query.match(/with\((.*?)\)\sand\s/);
```

Comment:
should we match at the beginning only, to further limit risk?

```suggestion
        const withClause = probe.query.match(/^with\((.*?)\)\sand\s/);
```